Procedural Geometry tutorial

clayjohn opened the pull request:
First step in #2076 

The other parts have been partially written for some time. I figured I would get this up for now to start the review process.

mhilbrunner commented on tutorials/3d/procedural_geometry/introduction.rst:
```diff
@@ -0,0 +1,106 @@
+Surfaces
+^^^^^^^^
+
+Each surface has its own material, alternatively, you can override the material for all surfaces
```
IMO, either "...material. Alternatively, ..." or adding an "or" inbetween ("...material, or alternatively...") sound better.

bitbutter replied:
There's a typo in index.rst: "procedural_geomerty/index"

mhilbrunner commented on tutorials/3d/procedural_geometry/introduction.rst:
```diff
@@ -0,0 +1,106 @@
+Surface array
+^^^^^^^^^^^^^
+
+The surface array has a is an array of length ``ArrayMesh.ARRAY_MAX``. Each position in the array is
```
Typo: **has a is an**

mhilbrunner commented on tutorials/3d/procedural_geometry/introduction.rst:
```diff
@@ -0,0 +1,106 @@
+The MeshDataTool is a resource that converts Mesh data into arrays of vertices, faces, and edges that can 
+be modified at runtime.
+
+For more information about the MeshDataTool, please see the `MeshDataTool tutorial` (coming soon).
```
This line should be removed. Just add the reference once the tutorial is actually there :)

mhilbrunner commented on tutorials/3d/procedural_geometry/introduction.rst:
```diff
@@ -0,0 +1,106 @@
+The SurfaceTool's primary purpose is to allow for the creation of Meshes using an OpenGL 1.x immediate mode style 
+interface. 
+
+For more information about the SurfaceTool, please see the `SurfaceTool and ImmediateGeometry tutorial` (coming soon).
```
Same as above. Of course, these may just be placeholders for your PR, as its still WIP... if so, please ignore me :P

mhilbrunner commented on tutorials/3d/procedural_geometry/introduction.rst:
```diff
@@ -0,0 +1,106 @@
+ImmediateGeometry is useful for prototyping because of the straightforward API, but it is slow becuase the geometry
+is rebuilt every frame. 
+
+For more information about ImmediateGeometry, please see the `SurfaceTool and ImmediateGeometry tutorial` (coming soon).
```
again, as above

mhilbrunner commented on tutorials/3d/procedural_geometry/introduction.rst:
```diff
@@ -0,0 +1,106 @@
+added to the scene tree and is drawn directly from the code. The SurfaceTool generates a Mesh that needs to be added 
+a MeshInstance to be seen. 
+
+ImmediateGeometry is useful for prototyping because of the straightforward API, but it is slow becuase the geometry
```
Typo: becuase

mhilbrunner commented:
This is going to be great :) Thanks for your continued efforts!

bitbutter commented on tutorials/3d/procedural_geometry/introduction.rst:
```diff
@@ -0,0 +1,106 @@
+-----------------
+
+Geometry is a fancy way of saying shape. In computer graphics, geometry is typically represented
+by an array of positions called "vertices". In Godot, geometry is represented by Meshes.
```
Isn't a mesh also composed of vertices?

clayjohn replied:
Kind of. Geometry is a math term for a 3D object while in graphics you usually represent geometry using a mesh. Which is an object with vertices, materials etc. In Godot it is slightly more complicated as a mesh is really just a container for surfaces, as explained below. So even though i have said geometry is represented by meshes, it would be inaccurate to say meshes are composed of vertices in godot. As meshes are a higher-level construct

follower commented on tutorials/3d/procedural_geometry/introduction.rst:
```diff
@@ -0,0 +1,106 @@
+a MultiMeshInstance. The MultiMeshInstance is used to draw meshes thousands of times very 
+cheaply. It takes advantage of hardware rendering in order to do so. The drawback with 
+using a MultiMeshInstance is that you are limited to one material for all instances. It uses an
+instance array to store different colors and transformations for each instance, but they all the
```
Suggestion: Delete leftover word: "but ~~they~~ all the instances" -> "but all the instances"

follower commented on tutorials/3d/procedural_geometry/introduction.rst:
```diff
@@ -0,0 +1,106 @@
+.. _doc_procedural_geometry_introduction:
+
+Introduction to procedural geometry
+===================================
```
I think this is a great topic for a tutorial. FWIW--in relation to reviewing the tutorial content--it might be helpful to know what audience you have in mind in terms of their experience (with Godot, with the concept of procedural generation, with geometry).

Based on the wording of "Geometry is a fancy way of saying shape. In computer graphics..." I'm picturing a reader who has little no experience with procedural generation nor computer-based geometry.

If my impression matches your intended target audience then it might be good to also include some examples of procedural generation (e.g. images and/or descriptions) and why someone might want to generate geometry in a procedural manner, rather than, say in 3D design software.

clayjohn replied:
Good idea. The idea was to capture a reader who has heard of procedural generation and wants to know what that means. We often get questions about whether or not Godot can "do procedural generation" because people see games like no mans sky. So this tutorial is meant to clear up what that means in a Godot-centered workflow

follower commented on tutorials/3d/procedural_geometry/introduction.rst:
```diff
@@ -0,0 +1,106 @@
+-----------------
+
+Geometry is a fancy way of saying shape. In computer graphics, geometry is typically represented
+by an array of positions called "vertices". In Godot, geometry is represented by Meshes.
```
If the target audience is beginner level it might be worth spelling out that:

 * A *position* refers to a point/location/position in 2D or 3D space. (Where 2D might be screen space and 2D/3D may be global/local "virtual" space.)  (e.g. (x,y) or (x,y,z) or (1,3) or (4, 2, 7) )
 * A point is also known as a "vertex" (with the plural "vertices" or "vertexes").

The non-obvious English language mapping from "vertex" to "vertices" I think contributes to making this topic more confusing for beginners than it actually needs to be. (I suspect this is also significant for people for whom English is not their first language.) (In addition to the fact I suspect many people encounter the concept of a "point" before knowing it as a "vertex".)

(Another point of potential confusion is how the word/concept of "Vector" does/doesn't fit into the terminology.)

Anyway, hope this feedback is helpful. It seems like a good idea to get feedback on WIP text like this while also keeping in mind that writing the complete 1st draft is the initial most important part before polishing. :) So my intention is to hopefully provide useful feedback in terms of audience-level/structure while also generally encouraging you to continue. :)

clayjohn replied:
While these are good considerations. I dont believe this is the correct place to get into the weeds on those subtle differences. By the time a user reaches this tutorial they will undoubtedly have had some experience and can be expected to know what a position is. I think it is enough to define "vertices" as an array of positions. 

Keep in mind the purpose if this tutorial series is to teach the user how to use arraymeshes, immediategeometry, surfacetool and meshdatatool. Not to serve as a comprehensive introduction to computer graphics.

follower commented:
FWIW having just read #2076 and seeing the term "Dynamic" rather than "Procedural" it makes me wonder about clearest term to use in the context of this tutorial.

I had been thinking you meant "Procedural" in the sense of [Procedural Generation](https://en.wikipedia.org/wiki/Procedural_generation) rather than in the sense of "dynamic"/programmatic drawing/construction.

Possible related/alternate terms: real-time generation; generated at run-time; immediate mode; constructed/drawn on the fly. Opposite to: pre-generated.

I feel like "Dynamic" is a more accurate term in relation to my understanding of your intended tutorial topic but I don't think it's a very meaningful term to a beginner/intro level audience. Might be good to see what term(s) are popular in other beginner level written/video tutorials on the topic. 

Words is hard, yo. :)

bitbutter commented:
Fwiw I think standardising to the 'procedural' prefix is correct with
respect to the way I'm familiar with the term. Meaning: using code (rather
than alternate means of sourcing, eg. recording the real world or designing
in an editor UI) to create geometry, audio, level layouts etc.

On Tue, Apr 16, 2019, 5:56 PM follower <notifications@github.com> wrote:

> FWIW having just read #2076
> <https://github.com/godotengine/godot-docs/issues/2076> and seeing the
> term "Dynamic" rather than "Procedural" it makes me wonder about clearest
> term to use in the context of this tutorial.
>
> I had been thinking you meant "Procedural" in the sense of Procedural
> Generation <https://en.wikipedia.org/wiki/Procedural_generation> rather
> than in the sense of "dynamic"/programmatic drawing/construction.
>
> Possible related/alternate terms: real-time generation; generated at
> run-time; immediate mode; constructed/drawn on the fly. Opposite to:
> pre-generated.
>
> I feel like "Dynamic" is a more accurate term in relation to my
> understanding of your intended tutorial topic but I don't think it's a very
> meaningful term to a beginner/intro level audience. Might be good to see
> what term(s) are popular in other beginner level written/video tutorials on
> the topic.
>
> Words is hard, yo. :)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot-docs/pull/2375#issuecomment-483722696>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADLSvszFfZLVZGouh80ZO_r9nksd6szks5vhfK0gaJpZM4cu-Hr>
> .
>

clayjohn commented:
While there are subtle differences between the terms. Procedural geometry is the best option. 

Procedural generation refers to the broad range of techniques used to generate things programmatically. Procedural geometry is geometry that has been procedurally generated. 

Dynamic geometry can be procedural geometry (and often would be), but is a more specific term that is usually reserved for things that change in real time. 

The issue was created in response to a patreon requested topic. I chose the title procedural geometry because it is both correct and more appropriate for the target audience.

mhilbrunner commented:
I'm also in favor of using 'procedural'.

follower commented:
> The issue was created in response to a patreon requested topic. I chose the title procedural geometry because it is both correct and more appropriate for the target audience.

Thanks for the additional context in this reply & https://github.com/godotengine/godot-docs/pull/2375#discussion_r275977287.

With that additional context, yes, I agree Procedural Geometry as a term accurately represents the topic you're covering.

akien-mga commented:
What's the status on this PR?

clayjohn commented:
@akien-mga I haven't forgotten about it, I have just had other priorities recently. I can post up the next section soon.

clayjohn commented:
Just pushed two changes.
1) added section on arraymesh
2) moved introduction into index.rst (the toc is now in the first section of introduction, it looks a little awkward, but it is necessary so the left hand menu works still)

clayjohn commented:
Added the ImmediateGeometry section. That should keep the reviewers busy. :D

clayjohn commented:
Just added the final two sections as well as some updates to index.rst!! Please review it everyone who is interested. Also, now would be a good time to add C# versions of the code. 

Once this is reviewed, I will rebase with the changes from master, move this to the "creating content" section and remove the small procedural geometry tutorial we currently have.

cbscribe commented on tutorials/content/procedural_geometry/meshdatatool.rst:
```diff
@@ -0,0 +1,119 @@
+The MeshDataTool is not used to generate geometry. But it is helpful for dynamically altering geometry, for example
+if you want to write a script to tessellate, simplify, or deform meshes. 
+
+The MeshDataTool is not as fast as altering arrays directly using ArrayMesh, however, it provides more information
```
Sentence break after ArrayMesh.

clayjohn commented:
@cbscribe This is great feedback, thanks! I will update the PR ASAP.

clayjohn commented:
@cbscribe Updated! :)

cbscribe commented:
Looks great. Let's get it out there so everyone can see it!

cbscribe commented:
@wisewoman547 This is not really the right place to ask for help. This repository is for submitting and discussing the documentation for Godot. If you need help or don't understand something, here are a few places you can get help:

* Godot QA: https://godotengine.org/qa
* Godot Discord:  https://discord.gg/zH7NUgz
* Godot Reddit: https://www.reddit.com/r/godot/

wisewoman547 commented:
Thanks,, but as previously stated I'm confused to ALL

On Aug 11, 2019 7:38 PM, "Chris Bradfield" <notifications@github.com> wrote:

@wisewoman547 <https://github.com/wisewoman547> This is not really the
right place to ask for help. This repository is for submitting and
discussing the documentation for Godot. If you need help or don't
understand something, here are a few places you can get help:

   - Godot QA: https://godotengine.org/qa
   - Godot Discord: https://discord.gg/zH7NUgz
   - Godot Reddit: https://www.reddit.com/r/godot/

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/godotengine/godot-docs/pull/2375?email_source=notifications&email_token=AM3375CWPYOXFPVEGFFT2ILQECPIXA5CNFSM4HF34HV2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4BLGAY#issuecomment-520270595>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AM3375CTFIYAHSAKZGOOMRDQECPIXANCNFSM4HF34HVQ>
.

cbscribe commented:
Then why are you posting on a closed discussion of a "Procedural Geometry" tutorial?

If you're new to Godot, this is the place to start: https://docs.godotengine.org/en/3.1/getting_started/step_by_step/intro_to_the_editor_interface.html
